File: VB/Controllers/HomeController.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.Web.Mvc
Imports System.Web.Script.Serialization
Imports TreeListOptimisticConcurrencyVBMvc.Models

Namespace TreeListOptimisticConcurrencyVBMvc.Controllers
    Public Class HomeController
        Inherits Controller
        Private db As New EmployeeDbContext()

        Public Function Index() As ActionResult
            Return View()
        End Function

        Public Function TreeListPartial() As ActionResult
            Return PartialView(db.Employees.ToList())
        End Function

        Public Function TreeListAddNewPartial(ByVal employee As Employee) As ActionResult
            Dim model = db.Employees

            If ModelState.IsValid Then
                Try
                    db.Entry(employee).State = System.Data.Entity.EntityState.Added
                    db.SaveChanges()
                Catch e As Exception
                    ViewData("EditError") = e.Message
                End Try
            Else
                ViewData("EditError") = "Please, correct all errors."
            End If

            Return PartialView("TreeListPartial", db.Employees.ToList())
        End Function

        Public Function TreeListUpdatePartial(ByVal employee As Employee) As ActionResult
            Dim model = db.Employees

            employee.RowVersion = CalculateOldRowVersion(employee.EmployeeID)

            If ModelState.IsValid Then
                Try
                    db.Entry(employee).State = System.Data.Entity.EntityState.Modified
                    db.SaveChanges()
                Catch e As Exception
                    ViewData("EditError") = e.Message
                End Try
            Else
                ViewData("EditError") = "Please, correct all errors."
            End If

            Return PartialView("TreeListPartial", db.Employees.ToList())
        End Function

        Public Function TreeListMovePartial(ByVal employeeID As Integer, ByVal supervisorID As Nullable(Of Integer)) As ActionResult
            Dim employee As Employee = db.Employees.Find(employeeID)

            If Not employee.SupervisorID.Equals(supervisorID) Then
                Try
                    db.Entry(employee).OriginalValues("RowVersion") = CalculateOldRowVersion(employeeID)
                    employee.SupervisorID = supervisorID
                    db.SaveChanges()
                Catch e As Exception
                    ViewData("EditError") = e.Message
                End Try
            End If

            Return PartialView("TreeListPartial", db.Employees.ToList())
        End Function

        Public Function TreeListDeletePartial(ByVal employee As Employee) As ActionResult
            Dim model = db.Employees

            employee.RowVersion = CalculateOldRowVersion(employee.EmployeeID)

            If ModelState.IsValid Then
                Try
                    db.Entry(employee).State = System.Data.Entity.EntityState.Deleted
                    db.SaveChanges()
                Catch e As Exception
                    ViewData("EditError") = e.Message
                End Try
            Else
                ViewData("EditError") = "Please, correct all errors."
            End If

            Return PartialView("TreeListPartial", db.Employees.ToList())
        End Function

        Private Function CalculateOldRowVersion(ByVal id As Integer) As Byte()
            Dim serializer As New JavaScriptSerializer()
            Dim rowVersions As String = Request("RowVersions")
            Dim dictionary As Dictionary(Of Object, String) = CType(serializer.Deserialize(rowVersions, GetType(Dictionary(Of Object, String))), Dictionary(Of Object, String))
            Dim rowVersion() As Char = dictionary(id.ToString()).ToCharArray()

            Return Convert.FromBase64CharArray(rowVersion, 0, rowVersion.Length)
        End Function
    End Class
End Namespace

File: CS/Controllers/HomeController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web.Mvc;
using System.Web.Script.Serialization;
using TreeListOptimisticConcurrencyMvc.Models;

namespace TreeListOptimisticConcurrencyMvc.Controllers {
    public class HomeController : Controller {
        private EmployeeDbContext db = new EmployeeDbContext();
        
        public ActionResult Index() {
            return View();
        }

        public ActionResult TreeListPartial() {
            return PartialView(db.Employees.ToList());
        }

        public ActionResult TreeListAddNewPartial(Employee employee) {
            var model = db.Employees;

            if (ModelState.IsValid) {
                try {
                    db.Entry(employee).State = System.Data.Entity.EntityState.Added;
                    db.SaveChanges();
                }
                catch (Exception e) {
                    ViewData["EditError"] = e.Message;
                }
            }
            else
                ViewData["EditError"] = "Please, correct all errors.";

            return PartialView("TreeListPartial", db.Employees.ToList());
        }

        public ActionResult TreeListUpdatePartial(Employee employee) {
            var model = db.Employees;

            employee.RowVersion = CalculateOldRowVersion(employee.EmployeeID);

            if (ModelState.IsValid) {
                try {
                    db.Entry(employee).State = System.Data.Entity.EntityState.Modified;
                    db.SaveChanges();
                }
                catch (Exception e) {
                    ViewData["EditError"] = e.Message;
                }
            }
            else
                ViewData["EditError"] = "Please, correct all errors.";

            return PartialView("TreeListPartial", db.Employees.ToList());
        }

        public ActionResult TreeListMovePartial(int employeeID, int? supervisorID) {
            Employee employee = db.Employees.Find(employeeID);

            if (employee.SupervisorID != supervisorID) {
                try {
                    db.Entry(employee).OriginalValues["RowVersion"] = CalculateOldRowVersion(employeeID);
                    employee.SupervisorID = supervisorID;
                    db.SaveChanges();
                }
                catch (Exception e) {
                    ViewData["EditError"] = e.Message;
                }
            }

            return PartialView("TreeListPartial", db.Employees.ToList());
        }

        public ActionResult TreeListDeletePartial(Employee employee) {
            var model = db.Employees;

            employee.RowVersion = CalculateOldRowVersion(employee.EmployeeID);

            if (ModelState.IsValid) {
                try {
                    db.Entry(employee).State = System.Data.Entity.EntityState.Deleted;
                    db.SaveChanges();
                }
                catch (Exception e) {
                    ViewData["EditError"] = e.Message;
                }
            }
            else
                ViewData["EditError"] = "Please, correct all errors.";

            return PartialView("TreeListPartial", db.Employees.ToList());
        }

        private byte[] CalculateOldRowVersion(int id) {
            JavaScriptSerializer serializer = new JavaScriptSerializer();
            string rowVersions = Request["RowVersions"];
            Dictionary<object, string> dictionary = (Dictionary<object, string>)serializer.Deserialize(rowVersions, typeof(Dictionary<object, string>));
            char[] rowVersion = dictionary[id.ToString()].ToCharArray();

            return Convert.FromBase64CharArray(rowVersion, 0, rowVersion.Length);
        }
    }
}